lib/python/treadmill_aws/ipakeytab.py

```diff
@@ -0,0 +1,172 @@
+from __future__ import division
+from __future__ import print_function
+from __future__ import unicode_literals
+
+import base64
+import io
+import logging
+import pwd
+import os
+import re
+import shutil
+import tempfile
+
+from twisted.internet import reactor
+from twisted.internet import protocol
+
+from treadmill import gssapiprotocol
+from treadmill import utils
+
+_LOGGER = logging.getLogger(__name__)
+
+
+def run_server(port, realm):
+    """Runs ipa keytab server."""
+    _LOGGER.info('IPA Keytab server starting - listening on port %d', port)
+
+    # no __init__ method.
+    #
+    # pylint: disable=W0232
+
+    def _error_str(string):
+        return str("ERROR: %s" % string).encode("utf-8")
+
+    class IPAKeytabServer(gssapiprotocol.GSSAPILineServer):
+        """IPA Keytab server."""
+
+        def __init__(self, realm):
+            self.realm = realm
+            gssapiprotocol.GSSAPILineServer.__init__(self)
+
+        def _validate_request(self, username):
+
+            if username == "admin" \
+                    or username == "host" \
+                    or username == "root":
+                raise ValueError(
+                    'keytabs for \'%s\' are not supported' %
+                    username)
+
+            try:
+                uid = pwd.getpwnam(username)
+            except KeyError:
+                raise KeyError("%s is not a valid username" % username)
+            except BaseException:
+                raise
+
+        def _validate_requestor(self, requestor):
+
+            match = re.search('^(.*)/(.*)@(.*)$', requestor)
+            if match is None:
+                raise ValueError(
+                    'Requestor %s must be a host service principal' %
+                    requestor)
+
+            requestor_name = match.group(1)
+            requestor_inst = match.group(2)
+            requestor_realm = match.group(3)
+
+            if requestor_name != "host":
+                raise ValueError(
+                    'Requestor %s is not a host principal' % requestor)
+
+            if self.realm is not None \
+                    and requestor_realm != self.realm:
+                raise ValueError(
+                    'Requestor \'%s\' must be in authorized realm \'%s\'' %
+                    (requestor, self.realm))
+
+            return (match.group(1), match.group(2), match.group(3))
+
+        def _get_keytab(self, name, inst, realm):
+            tmpktdir = tempfile.mkdtemp(prefix="ipakeytab-")
+            tmpkt = "%s/%s" % (tmpktdir, name)
+            service = "%s/%s@%s" % (name, inst, realm)
+
+            _LOGGER.info(
+                "writing keytab for %s/%s@%s => %s",
+                name,
+                inst,
+                realm,
+                tmpkt)
+
+            os.system(
+                "kadmin.local ktadd -k %s -norandkey %s" %
+                (tmpkt, service))
+            if not os.path.isfile(tmpkt):
+                os.system("klist -5 > /tmp/klist.out")
```

Comment:
we need to replace os.system with subproc.check_call.

it is separate PR, as it depends on aws-aliases and aws profile present, and this is still work in progress.